<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [3]:
import pandas as pd
import psycopg2

In [5]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [6]:
# текст запроса
query_3_1 = f'''
select
    count(id)
from
    public.vacancies
'''

In [7]:
# результат запроса
pd.read_sql_query(query_3_1, connection)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\1469115235.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [8]:
# текст запроса
query_3_2 = f'''
select
    count(id)
from
    public.employers
'''

In [9]:
# результат запроса
pd.read_sql_query(query_3_2, connection)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\4057088587.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [10]:
# текст запроса
query_3_3 = f'''
select
    count(id)
from
    public.areas
'''

In [11]:
# результат запроса
pd.read_sql_query(query_3_3, connection)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\4147122770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [12]:
# текст запроса
query_3_4 = f'''
select
    count(id)
from
    public.industries
'''

In [13]:
# результат запроса
pd.read_sql_query(query_3_4, connection)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\2627729322.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_4, connection)


count
0    294

***

## Выводы
По предварительному анализу данных можно сделать вывод, что существует давольно большое количество вакансий около 50 тыс.,
также стоит отметить, что количество работодателей в два раза меньше, их численность составляет 23 с половиной тысячи, следовательно,
можно скаазать, что позиция Data Scietist является востребованной. Более того ищут на эту позицию в 1362 регионах, что еще раз говорит
о ее востребованности, помимо всего Data Scientist-ы нужны в 294 сферах.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [14]:
# текст запроса
query_4_1 = f'''
select
    areas.name area,
    count(vacancies.id) cnt
from
    public.areas areas
    join public.vacancies vacancies on areas.id = vacancies.area_id
group by
    areas.name
order by
    cnt desc
'''

In [15]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
# выводим пятерку лидеров по числу вакансий среди регионов
df.head(25) 

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\2828769712.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0            Москва  5333
1   Санкт-Петербург  2851
2             Минск  2112
3       Новосибирск  2006
4            Алматы  1892
5      Екатеринбург  1698
6   Нижний Новгород  1670
7            Казань  1415
8         Краснодар  1301
9            Самара  1144
10   Ростов-на-Дону  1131
11          Воронеж  1063
12          Ташкент   851
13       Красноярск   847
14       Нур-Султан   840
15        Челябинск   786
16            Пермь   771
17              Уфа   767
18            Томск   692
19             Омск   617
20          Саратов   579
21           Тюмень   547
22      Владивосток   539
23          Тбилиси   474
24           Ижевск   466

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [16]:
# текст запроса
query_4_2 = f'''
select
    count(*)
from
    public.vacancies
where 
    salary_from is not null
    or salary_to is not null
'''

In [17]:
# результат запроса
pd.read_sql_query(query_4_2, connection)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\614600247.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [18]:
# текст запроса
query_4_3 = f'''
select
    avg(salary_from) salary_from_avg,
    avg(salary_to) salary_to_avg
from
    public.vacancies
where 
    salary_from is not null
    or salary_to is not null
'''

In [19]:
# результат запроса
pd.read_sql_query(query_4_3, connection)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\3534961570.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_3, connection)


salary_from_avg  salary_to_avg
0     71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [20]:
# текст запроса
query_4_4 = f'''
select
    schedule,
    employment,
    count(id)
from
    public.vacancies
group by
    schedule,
    employment
order by
    count desc
'''

In [21]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
# сочетание типов, которое назодится на втором месте по количеству вакансий
df.iloc[0]

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\112905211.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           Полный день
employment    Полная занятость
count                    35367
Name: 0, dtype: object

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [22]:
# текст запроса
query_4_5 = f'''
select
    experience,
    count(id) cnt
from
    public.vacancies
group by
    experience
order by
    cnt
'''

In [23]:
# результат запроса
pd.read_sql_query(query_4_5, connection)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\3190483664.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_5, connection)


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

## Выводы
Стоит заметить, что наибольшее количество вакансий находится в Москве более 5 тыс. и в Санкт-Петербурге около 3 тыс., в общем можно выделить около 10 городов, в которых существует большое количество вакансий(более 1000). Также примерно у половины вакансий заполнено хотя бы одно из полей с зарплатой, в среднем эти значения колеблятся от примерно 71 тыс. до 110 тыс. рублей. Чаще всего ищут людей на полный день и полную заннятость около 35 тыс., что состовляет больше половины вакансий. Наиболее востребованы люди с небольшим опытом работы от 1 года до 3 лет, дальше ищут людей со среднии опвтом работы, реже вакансии предлагают сотрудникам без опыта работы, и совсем редно нужны настоящие профессионалы с опытом раблоты более 6 лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [24]:
# текст запроса
query_5_1 = f'''
(select
    employers.name
from
    public.vacancies vacancies
    join public.employers employers on vacancies.employer_id = employers.id
group by
    employers.name
order by
    count(vacancies.id) desc
limit 1)
union all
(select
    employers.name
from
    public.vacancies vacancies
    join public.employers employers on vacancies.employer_id = employers.id
group by
    employers.name
order by
    count(vacancies.id) desc
offset 4
limit 1)
'''

In [25]:
# результат запроса
pd.read_sql_query(query_5_1, connection)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\3902369145.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_1, connection)


name
0         Яндекс
1  Газпром нефть

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [26]:
# текст запроса
query_5_2 = f'''
select
    areas.name area_name,
    count(employers.id) cnt_employers,
    count(vacancies.id) cnt_vacancies
from
    public.areas areas
    left join public.vacancies vacancies on vacancies.area_id = areas.id
    join public.employers employers on employers.area = areas.id
group by
    areas.name,
    areas.id
order by
    cnt_vacancies, cnt_employers desc
'''

In [27]:
df = pd.read_sql_query(query_5_2, connection)
# выводим первые пять строк
df.head(5)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\835431120.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


area_name  cnt_employers  cnt_vacancies
0              Россия            410              0
1           Казахстан            207              0
2  Московская область             75              0
3  Краснодарский край             19              0
4  Ростовская область             18              0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [28]:
# текст запроса
query_5_3 = f'''
select
    employers.name,
    count(distinct vacancies.area_id) area_count
from
    public.employers employers
    left join public.vacancies on employers.id = vacancies.employer_id
group by
    employers.name,
    employers.id
order by
    area_count desc
'''

In [29]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
# выводим первые пять строк
df.head(5)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\97296739.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  area_count
0                  Яндекс         181
1              Ростелеком         152
2              Спецремонт         116
3  Поляков Денис Иванович          88
4               ООО ЕФИН           71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [30]:
# текст запроса
query_5_4 = f'''
select
    count(*)
from
    public.employers employers
    left join public.employers_industries industries on employers.id = industries.employer_id
where
    industry_id is null
'''

In [31]:
# результат запроса
pd.read_sql_query(query_5_4, connection)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\1644714913.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [32]:
# текст запроса
query_5_5 = f'''
select
    employers.name
from
    public.employers employers
    left join public.employers_industries industries on employers.id = industries.employer_id
group by
    employers.id
having
    count(industries.industry_id) = 4
order by
    employers.name
offset 2
limit 1
'''

In [33]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
# выводим результат
display(df)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\387213333.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [34]:
# текст запроса
query_5_6 = f'''
select
    count(*)
from
    public.industries industries
    left join public.employers_industries employers_industries on industries.id = employers_industries.industry_id
where
    industries.name = 'Разработка программного обеспечения'
'''

In [35]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
# выводим результат
display(df)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\2764602813.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [36]:
# код для получения списка городов-милионников
import requests # импортируем библиотеку requests
from bs4 import BeautifulSoup # импортируем библиотеку BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России' # определяем значение URL страницы для запроса
response = requests.get(url) # выполняем GET-запрос, содержимое ответа присваивается переменной response
page = BeautifulSoup(response.text, 'html.parser') # создаём объект BeautifulSoup, указывая html-парсер
cities = page.find('table', class_ = 'standard').find_all('a') # получаем все ссылки в таблице содержащей города-милионники
cities = [city.text for city in cities[4:]] # получаем список городов-милионников
print(cities)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


In [37]:
# текст запроса
query_5_7 = f'''
(select
    areas.name,
    count(*) cnt
from
    public.vacancies vacancies
    join public.employers employers on employers.id = vacancies.employer_id
    join public.areas areas on areas.id = vacancies.area_id
where
    employers.name = 'Яндекс'
    and areas.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')
group by
    areas.name)
union all
(select
    'total',
    count(*)
from
    public.vacancies vacancies
    join public.employers employers on employers.id = vacancies.employer_id
    join public.areas areas on areas.id = vacancies.area_id
where
    employers.name = 'Яндекс'
    and areas.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')
)
order by cnt
'''

In [38]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
# выводим результат
display(df)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\2440688798.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            total  485

***

In [39]:
# выводы по анализу работодателей

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [40]:
# текст запроса
query_6_1 = f'''
select
    count(*)
from
    public.vacancies
where
    lower(name) like '%data%'
    or lower(name) like '%данн%'
'''

In [41]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
# выводим результат
display(df)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\2829793086.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [42]:
# текст запроса
query_6_2 = f'''
select
    count(*)
from
    public.vacancies
where
    (lower(name) like '%data scientist%'
    or lower(name) like '%data science%'
    or lower(name) like '%исследователь данных%'
    or (name like '%ML%' and name not like '%HTML%')
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%')
    and 
    (lower(name) like '%junior%' 
    or experience = 'Нет опыта'
    or employment = 'Стажировка')
'''

In [43]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
# выводим результат
display(df)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\2444402593.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [44]:
# текст запроса
query_6_3 = f'''
select
    count(*)
from
    public.vacancies
where
    (lower(name) like '%data scientist%'
    or lower(name) like '%data science%'
    or lower(name) like '%исследователь данных%'
    or (name like '%ML%' and name not like '%HTML%')
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%')
    and 
    (lower(key_skills) like '%sql%'
    or lower(key_skills) like '%postgres%')
'''

In [45]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
# выводим результат
display(df)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\147728065.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [46]:
# текст запроса
query_6_4 = f'''
select
    count(*)
from
    public.vacancies
where
    (lower(name) like '%data scientist%'
    or lower(name) like '%data science%'
    or lower(name) like '%исследователь данных%'
    or (name like '%ML%' and name not like '%HTML%')
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%')
    and 
    (lower(key_skills) like '%python%')
'''

In [47]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
# выводим результат
display(df)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\2384938394.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [60]:
# текст запроса
query_6_5 = f'''
select
    key_skills
from
    public.vacancies
where
    (lower(name) like '%data scientist%'
    or lower(name) like '%data science%'
    or lower(name) like '%исследователь данных%'
    or (name like '%ML%' and name not like '%HTML%')
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%')
'''

In [66]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
print(df.iloc[0])
df['skills_count'] = df['key_skills'].apply(lambda x: len(str(x).split('\t')))
avg = round(df['skills_count'].mean(), 2)
# выводим результат
print(df.iloc[0])
print(avg)

C:\Users\drfre\AppData\Local\Temp\ipykernel_20388\493381807.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


key_skills    Python\tSQL\tРабота в команде\tOpenCV\tML
Name: 0, dtype: object
key_skills      Python\tSQL\tРабота в команде\tOpenCV\tML
skills_count                                            5
Name: 0, dtype: object
5.85


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [50]:
# текст запроса

In [51]:
# результат запроса

***

In [52]:
# выводы по предметному анализу

# Общий вывод по проекту

In [53]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования